From 5a51a257214d8800cc9d82b3e58328a8b29786b0 Mon Sep 17 00:00:00 2001
From: Julian Bouzas <julian.bouzas@collabora.com>
Date: Sat, 25 Jul 2026 16:10:06 -0400
Subject: [PATCH] spa: alsa: pass ELD data to ACP HDMI profile recheck

Reuse the ELD data that is already parsed in the HDMI ELD mixer callback when
rechecking ACP HDMI profile support, instead of reparsing channel information
from /proc.
Origin: upstream, https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/5a51a257214d8800cc9d82b3e58328a8b29786b0
Forwarded: not-needed
Last-Update: 2026-10-13
---
 spa/plugins/alsa/acp/acp.c        |  2 +-
 spa/plugins/alsa/acp/alsa-mixer.c | 65 ++-----------------------------
 spa/plugins/alsa/acp/alsa-mixer.h |  2 +-
 spa/plugins/alsa/acp/alsa-util.h  | 20 +++++-----
 4 files changed, 16 insertions(+), 73 deletions(-)

diff --git a/spa/plugins/alsa/acp/acp.c b/spa/plugins/alsa/acp/acp.c
index 8df0fb7fd..651b27c2e 100644
--- a/spa/plugins/alsa/acp/acp.c
+++ b/spa/plugins/alsa/acp/acp.c
@@ -1219,7 +1219,7 @@ static int hdmi_eld_changed(snd_mixer_elem_t *melem, unsigned int mask)
 		impl->events->props_changed(impl->user_data);
 
 	if (changed && mask != 0) {
-		pa_alsa_profile_set_recheck_hdmi_eld(impl->profile_set, impl->card.index);
+		pa_alsa_profile_set_recheck_hdmi_eld(impl->profile_set, impl->card.index, &eld);
 		update_profile_availabilities(impl);
 	}
 
diff --git a/spa/plugins/alsa/acp/alsa-mixer.c b/spa/plugins/alsa/acp/alsa-mixer.c
index eff00bf0e..afea1d9e9 100644
--- a/spa/plugins/alsa/acp/alsa-mixer.c
+++ b/spa/plugins/alsa/acp/alsa-mixer.c
@@ -20,7 +20,6 @@
 
 #include "config.h"
 
-#include <dirent.h>
 #include <sys/types.h>
 #include <alsa/asoundlib.h>
 #include <math.h>
@@ -5276,64 +5275,10 @@ void pa_alsa_profile_set_probe(
     ps->probed = true;
 }
 
-static int get_hdmi_eld_lpcm_channels(uint32_t card_index)
-{
-    char dir_path[256];
-    DIR *dir;
-    struct dirent *de;
-
-    snprintf(dir_path, sizeof(dir_path), "/proc/asound/card%u", card_index);
-
-    if ((dir = opendir(dir_path)) == NULL)
-        return -1;
-
-    while ((de = readdir(dir)) != NULL) {
-        int card, eld_idx;
-        char path[512];
-        FILE *f;
-        char line[256];
-        int monitor_present = 0, eld_valid = 0;
-        int lpcm_channels = 0;
-        bool next_channels_is_lpcm = false;
-
-        if (sscanf(de->d_name, "eld#%d.%d", &card, &eld_idx) != 2)
-            continue;
-
-        snprintf(path, sizeof(path), "%s/%s", dir_path, de->d_name);
-        if ((f = fopen(path, "r")) == NULL)
-            continue;
-
-        while (fgets(line, sizeof(line), f)) {
-            int val;
-
-            if (sscanf(line, "monitor_present %d", &val) == 1)
-                monitor_present = val;
-            else if (sscanf(line, "eld_valid %d", &val) == 1)
-                eld_valid = val;
-            else if (strstr(line, "_coding_type"))
-                next_channels_is_lpcm = strstr(line, "LPCM") != NULL;
-            else if (next_channels_is_lpcm &&
-                     sscanf(line, "sad%*u_channels %d", &val) == 1) {
-                lpcm_channels = val;
-		break;
-            }
-        }
-
-        fclose(f);
-
-        if (monitor_present && eld_valid) {
-            closedir(dir);
-            return lpcm_channels;
-        }
-    }
-
-    closedir(dir);
-    return -1;
-}
-
 void pa_alsa_profile_set_recheck_hdmi_eld(
         pa_alsa_profile_set *ps,
-        uint32_t card_index) {
+        uint32_t card_index,
+        const struct pa_hdmi_eld *eld) {
 
     pa_alsa_profile *p;
     void *state;
@@ -5359,14 +5304,12 @@ void pa_alsa_profile_set_recheck_hdmi_eld(
 
         if (p->output_mappings) {
             PA_IDXSET_FOREACH(m, p->output_mappings, idx) {
-                int lpcm_channels;
 
-		/* Only handle HDMI mappings */
+                /* Only handle HDMI mappings */
                 if (strcmp (m->device_strings[0], "hdmi:%f") != 0)
                     continue;
 
-                lpcm_channels = get_hdmi_eld_lpcm_channels(card_index);
-                if (lpcm_channels < (int)m->channel_map.channels) {
+                if (eld->lpcm_channels < (int)m->channel_map.channels) {
                     p->supported = false;
                     pa_log_debug("HDMI profile %s disabled (ELD channels did not match)", p->name);
                     break;
diff --git a/spa/plugins/alsa/acp/alsa-mixer.h b/spa/plugins/alsa/acp/alsa-mixer.h
index 9da453fa7..7945b0432 100644
--- a/spa/plugins/alsa/acp/alsa-mixer.h
+++ b/spa/plugins/alsa/acp/alsa-mixer.h
@@ -432,7 +432,7 @@ void pa_alsa_profile_free (pa_alsa_profile *p);
 
 pa_alsa_profile_set* pa_alsa_profile_set_new(const char *fname, const pa_channel_map *bonus);
 void pa_alsa_profile_set_probe(pa_alsa_profile_set *ps, pa_hashmap *mixers, uint32_t card_index, const pa_sample_spec *ss, unsigned default_n_fragments, unsigned default_fragment_size_msec);
-void pa_alsa_profile_set_recheck_hdmi_eld(pa_alsa_profile_set *ps, uint32_t card_index);
+void pa_alsa_profile_set_recheck_hdmi_eld(pa_alsa_profile_set *ps, uint32_t card_index, const struct pa_hdmi_eld *eld);
 void pa_alsa_profile_set_free(pa_alsa_profile_set *s);
 void pa_alsa_profile_set_dump(pa_alsa_profile_set *s);
 void pa_alsa_profile_set_drop_unsupported(pa_alsa_profile_set *s);
diff --git a/spa/plugins/alsa/acp/alsa-util.h b/spa/plugins/alsa/acp/alsa-util.h
index 87e463c42..ec1426ddb 100644
--- a/spa/plugins/alsa/acp/alsa-util.h
+++ b/spa/plugins/alsa/acp/alsa-util.h
@@ -27,6 +27,16 @@
 
 #include "compat.h"
 
+typedef struct pa_hdmi_eld pa_hdmi_eld;
+struct pa_hdmi_eld {
+    char monitor_name[17];
+    uint8_t speakers;
+    uint64_t iec958_codecs;
+    uint8_t lpcm_channels;
+};
+
+int pa_alsa_get_hdmi_eld(snd_hctl_elem_t *elem, pa_hdmi_eld *eld);
+
 #include "alsa-mixer.h"
 
 enum {
@@ -172,14 +182,4 @@ void pa_alsa_mixer_set_fdlist(pa_hashmap *mixers, snd_mixer_t *mixer, pa_mainloo
 #endif
 void pa_alsa_mixer_free(pa_alsa_mixer *mixer);
 
-typedef struct pa_hdmi_eld pa_hdmi_eld;
-struct pa_hdmi_eld {
-    char monitor_name[17];
-    uint8_t speakers;
-    uint64_t iec958_codecs;
-    uint8_t lpcm_channels;
-};
-
-int pa_alsa_get_hdmi_eld(snd_hctl_elem_t *elem, pa_hdmi_eld *eld);
-
 #endif
